src/directdb.rs

```diff
@@ -0,0 +1,146 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright 2020 Joyent, Inc.
+ */
+
+use crossbeam_channel as crossbeam;
+use futures::{pin_mut, TryStreamExt};
+use serde_json::{self, Value};
+use std::io::{Error, ErrorKind};
+use tokio_postgres::NoTls;
+
+use crate::config::Config;
+use crate::{
+    get_sharks_from_manta_obj, object_id_from_manta_obj, SharkspotterMessage,
+};
+use serde::{Deserialize, Serialize};
+use slog::{debug, error, trace, warn, Logger};
+
+// Unfortunately the Manta records in the moray database are slightly
+// different from what we get back from the moray service (both for the
+// `findobjects` and `sql` endpoints.  So if we are going direct to the database
+// we need to use a different struct to represent the record (DB schema).
+//
+// moray=> SELECT table_name, column_name, data_type FROM information_schema.columns WHERE table_name = 'manta';
+// table_name | column_name | data_type
+// ------------+-------------+-----------
+//  manta      | _id         | bigint
+//  manta      | _txn_snap   | integer
+//  manta      | _key        | text
+//  manta      | _value      | text
+//  manta      | _etag       | character
+//  manta      | _mtime      | bigint
+//  manta      | _vnode      | bigint
+//  manta      | dirname     | text
+//  manta      | name        | text
+//  manta      | owner       | text
+//  manta      | objectid    | text
+//  manta      | type        | text
+#[derive(Deserialize, Serialize)]
+struct MorayMantaBucketObject {
+    _id: i64,
+    _txn_snap: Option<i32>,
+    _key: String,
+    _value: String,
+    _etag: String,
+    _mtime: i64,
+    _vnode: i64,
+    dirname: String,
+    name: String,
+    owner: String,
+    objectid: String,
+    #[serde(alias = "type")]
+    record_type: String,
+}
+
+pub async fn get_objects_from_shard(
+    shard: u32,
+    conf: Config,
+    log: Logger,
+    obj_tx: crossbeam::Sender<SharkspotterMessage>,
+) -> Result<(), Error> {
+    let shard_host_name =
+        format!("{}.rebalancer-postgres.{}", shard, conf.domain);
+
+    debug!(log, "Connecting to {}", shard_host_name);
+    let (client, connection) = tokio_postgres::Config::new()
```

Comment:
Had to go through the documentation to figure out what this was doing.  I'm consulting [the tokio_postgres 0.5.5 documentation](https://docs.rs/tokio-postgres/0.5.5/tokio_postgres/) as I go through.  A little bit of commentary around what's happening would be helpful.

Reply:
added some comments.